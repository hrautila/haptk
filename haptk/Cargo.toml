[package]
name = "haptk"
version = "0.3.0"
authors = ["Osma S. Rautila"]
edition = "2021"
description = "HAPTK - Haplotype analysis toolkit"
license = "GPL-3.0-only"
readme = "README.md"
homepage = "https://github.com/xosxos/haptk"
repository = "https://github.com/xosxos/haptk"
exclude = [
    "tests/*",
]


[lib]
crate-type = ["cdylib", "rlib"]
path = "src/lib.rs"
doctest = false

[[bin]]
name = "haptk"
path = "src/bin.rs"

[features]
default = ["clap"]
experimental = ["clap", "dep:fishers_exact"]
clap = ["dep:clap", "dep:anstyle"]

[dependencies]
alphanumeric-sort = "1.5"
bgzip = "0.3"
bio = "2.0"
niffler = "2.6"
ndarray = { version = "0.16", features = ["rayon", "blas", "serde"] }
ndarray-npy = { version = "0.9", default-features = false }
termion = "4.0"
color-eyre = "0.6"
rayon = "1.10"
serde = { version = "1.0", features = ["derive"] }
rand = "0.8.5"
svg = "0.13"
csv = "1.3"
itertools = "0.13"
serde_json = "1.0"
tracing = "0.1"
tracing-appender = "0.2"
indexmap = {version = "2.6", features = ["serde"]}
tracing-subscriber = { version = "0.3", default-features = false, features = ["std", "fmt", "time"] }
time = { version = "0.3", default-features = false, features = ["local-offset"] }
petgraph = { version = "0.6", default-features = false, features = ["serde-1"] }
rust-htslib = { version = "0.44", default-features = false, features = ['bzip2'] }
polars = { version = "=0.33.2", default-features = false, features = ["lazy", "is_in", "csv", "fmt", "serde"]}
image = { version = "0.25", default-features = false, features = ['png'] }
statrs = { version = "0.16", default-features = false }
chrono = { version = "0.4.38" }

# Cli
clap = { version = "=4.4.18", optional = true, features = ["derive"] }
anstyle = { version = "1.0.7", optional = true }

# Genome-wide
fishers_exact = { version="1.0.1", optional=true }


[dev-dependencies]
insta = { version = "1.26.0", default-features = false, features = ["yaml"] }



