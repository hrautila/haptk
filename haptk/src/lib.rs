// HATK - Haplotype analysis toolkit
// Copyright (C) 2022  Osma S. Rautila
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

pub mod libs;
pub use libs::{args, error, io, read_vcf, stats, structs, utils};

#[cfg(feature = "clap")]
pub use libs::clap;

/// Graphs implemented in Rust, HST graph will be replaced by the Python implementation
pub mod graphs;

/// HAPTK commands
pub mod subcommands;
