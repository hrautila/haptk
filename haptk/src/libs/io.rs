use std::collections::BTreeMap;

use std::collections::HashMap;
use std::ffi::{OsStr, OsString};
use std::fs::File;
use std::io::{self, BufRead};
use std::path::{Path, PathBuf};
use std::sync::mpsc::Receiver;
use std::thread::{self, JoinHandle};

use bgzip::tabix::Tabix;
use color_eyre::eyre::{eyre, OptionExt, WrapErr};
use color_eyre::Result;
use csv::{QuoteStyle, Reader, ReaderBuilder, Writer, WriterBuilder};
use rust_htslib::bcf::{header::HeaderRecord, IndexedReader, Read};

use crate::args::{Selection, StandardArgs};
use crate::read_vcf::get_reader;
use crate::structs::Coord;
use crate::structs::HapVariant;
use crate::utils::strip_prefix;

#[cfg(feature = "experimental")]
use crate::subcommands::scan::scan_segregate::CoordSamples;

#[allow(clippy::upper_case_acronyms)]
pub enum FileType {
    BED,
    VCF,
    CSV,
    TSV,
    HST,
}

impl FileType {
    pub fn from_path(path: &PathBuf) -> Result<Self> {
        let extension: &str = Path::new(&path)
            .extension()
            .and_then(OsStr::to_str)
            .ok_or_else(|| eyre!("No filetype in path"))?;

        let extension = match extension {
            "gz" | "bgz" => return_double_extension_filetype(path, extension)?,
            _ => extension.to_string(),
        };

        Ok(match extension.as_str() {
            "vcf.gz" | "vcf" | "bcf" => Self::VCF,
            "hst.gz" | "hst" => Self::HST,
            "bed.gz" | "bed" => Self::BED,
            "csv" | "ids" => Self::CSV,
            "tsv" => Self::TSV,
            _ => return Err(eyre!("File extension: {extension} is not supported")),
        })
    }
}

pub fn return_double_extension_filetype(path: &Path, e1: &str) -> Result<String> {
    let stem = path
        .file_stem()
        .and_then(OsStr::to_str)
        .ok_or_else(|| eyre!("file has no stem"))?;
    let e2 = Path::new(&stem)
        .extension()
        .and_then(OsStr::to_str)
        .ok_or_else(|| eyre!("file has no other filetype"))?;
    Ok(format!("{e2}.{e1}"))
}

pub fn read_coords_file(path: &PathBuf) -> Result<Vec<Coord>> {
    let input = get_input(Some(path.clone()))?;

    let mut rdr = match FileType::from_path(path)? {
        FileType::CSV => get_csv_reader(input),
        FileType::TSV => get_tsv_reader(input, false),
        FileType::BED => get_tsv_reader(input, false),
        _ => return Err(eyre!("Filetype from {path:?} is not supported")),
    };

    let mut variants: Vec<Coord> = vec![];

    for line in rdr.records() {
        let record = line?;
        let variant: Coord = record.deserialize(None)?;
        variants.push(variant);
    }

    Ok(variants)
}

#[cfg(feature = "experimental")]
pub fn read_coords_sample_file(path: &PathBuf) -> Result<Vec<CoordSamples>> {
    let input = get_input(Some(path.clone()))?;

    let mut rdr = match FileType::from_path(path)? {
        FileType::CSV => get_csv_reader(input),
        FileType::TSV => get_tsv_reader(input, false),
        FileType::BED => get_tsv_reader(input, false),
        _ => return Err(eyre!("Filetype from {path:?} is not supported")),
    };

    let mut variants: Vec<CoordSamples> = vec![];

    for line in rdr.records() {
        let record = line?;
        let variant: CoordSamples = record.deserialize(None)?;
        variants.push(variant);
    }

    Ok(variants)
}

pub fn filter_out_non_vcf_sample_names(
    path: &PathBuf,
    contig: &str,
    wanted_samples: Vec<String>,
) -> Result<Vec<String>> {
    let reader = get_reader(path, contig, None)?;

    let samples = reader
        .header()
        .samples()
        .into_iter()
        .map(|sample| Ok(String::from_utf8_lossy(sample).to_string()))
        .collect::<Result<Vec<String>>>()?;

    for i in &wanted_samples {
        if !samples.contains(i) {
            tracing::warn!("Wanted sample {i} is not in the VCF");
        }
    }

    Ok(samples
        .into_iter()
        .filter(|s| wanted_samples.contains(s))
        .collect())
}

#[derive(Debug, Clone, serde::Deserialize)]
struct RecombinationRow<'a> {
    chr: &'a str,
    pos: u64,
    _rate: f32,
    cm: f32,
}

pub fn check_recombination_file_matches_contig(path: PathBuf, contig: &str) -> Result<bool> {
    let mut rdr = get_tsv_reader(get_input(Some(path))?, false);
    let mut lines = rdr.records();

    let line = lines
        .next()
        .ok_or_eyre("Recombination rates file {path} contains 0 rows.")?;

    let record = line?;
    let row: RecombinationRow = record.deserialize(None).
            wrap_err(eyre!("Make sure no headers are present and that the recombination file is in order chr,pos,rate,cm. If the issue is not fixed, you have an invalid field in the file."))?;
    Ok(row.chr == contig)
}

pub fn read_recombination_file(path: PathBuf) -> Result<BTreeMap<u64, f32>> {
    let mut rdr = get_tsv_reader(get_input(Some(path))?, false);
    let mut rates = BTreeMap::new();

    let mut last_cm = 0.0;
    let mut last_pos = 0;
    for line in rdr.records() {
        let record = line?;
        let row: RecombinationRow = record.deserialize(None).
            wrap_err(eyre!("Make sure no headers are present and that the recombination file is in order chr,pos,rate,cm. If the issue is not fixed, you have an invalid field in the file."))?;
        if last_cm > row.cm {
            return Err(eyre!(
                "Recombination rates file is not sorted for centimorgans"
            ));
        }
        if last_pos > row.pos {
            return Err(eyre!(
                "Recombination rates file is not sorted for positions"
            ));
        }
        last_cm = row.cm;
        last_pos = row.pos;
        rates.insert(row.pos, row.cm);
    }

    Ok(rates)
}

pub fn read_haplotype_file(ht_path: PathBuf) -> Result<Vec<HapVariant>> {
    let input = get_input(Some(ht_path))?;
    let mut rdr = get_csv_reader(input);

    let mut variants: Vec<HapVariant> = vec![];

    for line in rdr.records() {
        let record = line?;
        let variant: HapVariant = record.deserialize(None)?;
        if let Some(latest) = variants.last() {
            if latest.pos > variant.pos {
                return Err(eyre!(
                    "The haplotype file is not sorted by position. {} is larger than {}",
                    latest,
                    variant
                ));
            }
        }
        variants.push(variant);
    }

    Ok(variants)
}

pub fn read_lines<P>(filename: P) -> Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    // let file = File::open(filename)?;
    let name = filename.as_ref().display();
    let file = match File::open(&filename) {
        Ok(x) => x,
        Err(err) => {
            let msg = format!("failed to open {name}: {err}");
            return Err(std::io::Error::new(std::io::ErrorKind::NotFound, msg))?;
        }
    };
    Ok(io::BufReader::new(file).lines())
}

pub fn read_multiple_sample_ids(path: &Option<Vec<PathBuf>>) -> Result<Option<Vec<String>>> {
    match path {
        Some(paths) => {
            let mut samples = vec![];
            for path in paths {
                for line in read_lines(path)?.map_while(Result::ok) {
                    let line = line.trim();
                    samples.push(line.to_string());
                }
            }
            Ok(Some(samples))
        }
        None => Ok(None),
    }
}

pub fn read_sample_ids(path: &Option<PathBuf>) -> Result<Option<Vec<String>>> {
    match path {
        Some(path) => {
            let mut samples = vec![];

            for line in read_lines(path)?.map_while(Result::ok) {
                let line = line.trim();
                samples.push(line.to_string());
            }
            Ok(Some(samples))
        }
        None => Ok(None),
    }
}

pub fn push_to_output(args: &StandardArgs, output: &mut PathBuf, name: &str, suffix: &str) {
    if let Some(prefix) = &strip_prefix(args.prefix.clone()) {
        match args.selection {
            Selection::All => output.push(format!("{prefix}_{name}.{suffix}")),
            Selection::OnlyAlts => output.push(format!("{prefix}_{name}_only_alts.{suffix}")),
            Selection::OnlyRefs => output.push(format!("{prefix}_{name}_only_refs.{suffix}")),
            Selection::OnlyLongest => output.push(format!("{prefix}_{name}_only_longest.{suffix}")),
            Selection::Unphased => output.push(format!("{prefix}_{name}_rwc.{suffix}")),
            Selection::Haploid => output.push(format!("{prefix}_{name}_haploid.{suffix}")),
        }
    } else {
        match args.selection {
            Selection::All => output.push(format!("{name}.{suffix}")),
            Selection::OnlyAlts => output.push(format!("{name}_only_alts.{suffix}")),
            Selection::OnlyRefs => output.push(format!("{name}_only_refs.{suffix}")),
            Selection::OnlyLongest => output.push(format!("{name}_only_longest.{suffix}")),
            Selection::Unphased => output.push(format!("{name}_rwc.{suffix}")),
            Selection::Haploid => output.push(format!("{name}_haploid.{suffix}")),
        }
    }
}

pub fn get_htslib_contig_len(path: &PathBuf, contig: &str) -> Result<u64> {
    for r in IndexedReader::from_path(path)?
        .header()
        .header_records()
        .iter()
        .filter(|r| matches!(r, HeaderRecord::Contig { .. }))
    {
        match r {
            HeaderRecord::Contig { values, .. } => {
                if let Some(id) = values.get("ID") {
                    if id == contig {
                        return Ok(values
                            .get("length")
                            .ok_or_else(|| eyre!("VCF header has no contig length for {id}"))?
                            .parse::<u64>()?);
                    }
                }
            }
            _ => {
                panic!("Header contig filtering failed for some reason. Create an issue at Github")
            }
        }
    }
    Err(eyre!(
        "Cannot get length for contig {contig}. Check the vcf header."
    ))
}

pub fn get_htslib_bcf_contigs(path: &PathBuf) -> Result<Vec<(String, i64)>> {
    IndexedReader::from_path(path)?
        .header()
        .header_records()
        .iter()
        .filter(|r| matches!(r, HeaderRecord::Contig { .. }))
        .map(|r| match r {
            HeaderRecord::Contig { values, .. } => {
                let id = values
                    .get("ID")
                    .ok_or_eyre("The input VCF has no ID for some contig record in the header")?;
                Ok((
                    id.clone(),
                    values
                        .get("length")
                        .ok_or_eyre("VCF header has no contig length for {id}")?
                        .parse::<i64>()?,
                ))
            }
            _ => {
                panic!("Header contig filtering failed for some reason. Create an issue at Github")
            }
        })
        .collect()
}

pub fn get_tsv_reader<R: io::Read>(input: R, has_headers: bool) -> Reader<R> {
    ReaderBuilder::new()
        .delimiter(b'\t')
        .has_headers(has_headers)
        .flexible(false)
        .from_reader(input)
}

pub fn get_csv_reader<R: io::Read>(input: R) -> Reader<R> {
    ReaderBuilder::new()
        .delimiter(b',')
        .has_headers(true)
        .flexible(false)
        .from_reader(input)
}

pub fn get_csv_writer<W: io::Write>(output: W) -> Writer<W> {
    WriterBuilder::new()
        .delimiter(b',')
        .has_headers(false)
        .flexible(true)
        .from_writer(output)
}

pub fn get_strict_tsv_writer<W: io::Write>(output: W) -> Writer<W> {
    WriterBuilder::new()
        .delimiter(b'\t')
        .has_headers(false)
        .flexible(true)
        .from_writer(output)
}

pub fn get_vcf_writer<W: io::Write>(output: W) -> Writer<W> {
    WriterBuilder::new()
        .delimiter(b'\t')
        .has_headers(false)
        .flexible(true)
        .double_quote(false)
        .quote_style(QuoteStyle::Never)
        .from_writer(output)
}

pub fn get_input(filename: Option<PathBuf>) -> Result<Box<dyn io::Read>> {
    let input: Box<dyn io::Read> = match filename {
        Some(name) => match name.to_str() {
            Some("-") => Box::new(io::stdin()),
            Some(name) => {
                let r = match niffler::from_path(name) {
                    Ok(x) => x.0,
                    Err(err) => {
                        let msg = format!("failed to open \"{name}\": {err}");
                        return Err(eyre!(msg))?;
                    }
                };
                Box::new(r)
            }
            None => return Err(eyre!("Unknown I/O error")),
        },
        None => Box::new(io::stdin()),
    };
    Ok(input)
}

pub fn get_output(filename: Option<PathBuf>) -> Result<Box<dyn io::Write>> {
    let output: Box<dyn io::Write> = match filename {
        Some(name) => match name.to_str() {
            Some("-") => Box::new(io::stdout()),
            Some(name) => Box::new(
                match std::fs::File::options()
                    .create(true)
                    .write(true)
                    .truncate(true)
                    .open(name)
                {
                    Ok(x) => x,
                    Err(err) => return Err(eyre!("failed to open \"{name}\": {err}"))?,
                },
            ),
            None => return Err(eyre!("Unknown I/O error")),
        },
        None => Box::new(io::stdout()),
    };
    Ok(output)
}

pub fn open_csv_writer(name: PathBuf) -> Result<Writer<Box<dyn io::Write>>> {
    Ok(get_csv_writer(get_output(Some(name))?))
}

pub fn open_strict_tsv_writer(name: PathBuf) -> Result<Writer<Box<dyn io::Write>>> {
    Ok(get_strict_tsv_writer(get_output(Some(name))?))
}

pub fn spawn_csv_collector(
    rx: Receiver<Vec<String>>,
    header: Vec<String>,
    outname: Option<PathBuf>,
    sort: bool,
) -> JoinHandle<Result<()>> {
    thread::spawn(move || -> Result<()> {
        let output = get_output(outname)?;
        let mut output = get_csv_writer(output);
        output.write_record(header)?;

        match sort {
            true => {
                let mut lines = vec![];

                while let Ok(line) = rx.recv() {
                    lines.push(line);
                }

                lines.sort_by(|a, b| alphanumeric_sort::compare_str(&a[0], &b[0]));

                for line in lines {
                    output.write_record(&line)?;
                }
            }
            false => {
                while let Ok(line) = rx.recv() {
                    output.write_record(&line)?;
                }
            }
        }

        Ok(())
    })
}

pub fn append_ext(ext: impl AsRef<OsStr>, path: &PathBuf) -> PathBuf {
    let mut os_string: OsString = path.into();
    os_string.push(".");
    os_string.push(ext.as_ref());
    os_string.into()
}

pub fn read_tabix(path: &PathBuf) -> Result<HashMap<String, (u64, usize)>> {
    let path = append_ext("tbi", path);
    let mut file = match std::fs::File::open(&path) {
        Ok(x) => x,
        Err(err) => {
            let msg = format!("failed to open \"{}\": {err}", path.display());
            return Err(eyre!(msg))?;
        }
    };
    let tabix = Tabix::from_reader(&mut file)?;

    let mut contigs: HashMap<String, (u64, usize)> = HashMap::new();

    let names = tabix
        .names
        .iter()
        .map(|n| Ok(std::str::from_utf8(n)?))
        .collect::<Result<Vec<&str>>>()?;

    let start = tabix
        .sequences
        .iter()
        .map(|n| {
            Ok((
                *n.intervals.first().unwrap(),
                n.number_of_distinct_bin as usize,
            ))
        })
        .collect::<Result<Vec<(u64, usize)>>>()?;

    names.iter().zip(start).for_each(|(name, (start, end))| {
        let _ = contigs.insert(name.trim_matches(char::from(0)).to_string(), (start, end));
    });

    Ok(contigs)
}

pub fn write_haplotype(
    haplotype: Vec<HapVariant>,
    mut writer: csv::Writer<Box<dyn std::io::Write>>,
) -> Result<()> {
    writer.write_record(vec!["contig", "pos", "ref", "alt", "gt"])?;

    for coord in haplotype {
        writer.write_record(vec![
            coord.contig,
            coord.pos.to_string(),
            coord.reference,
            coord.alt,
            coord.gt.to_string(),
        ])?;
    }
    Ok(())
}

#[cfg(test)]
#[rustfmt::skip]
mod tests {
    use super::*;

    #[test]
    fn test_push_to_output() {
        let mut output = std::path::PathBuf::new();
        let args = crate::args::StandardArgs::default();
        push_to_output(&args, &mut output, "picture", "png");
        assert_eq!(output, std::path::PathBuf::from("picture.png"));

        let mut output = std::path::PathBuf::from("./foo");
        let args = crate::args::StandardArgs::default();
        push_to_output(&args, &mut output, "picture", "png");
        assert_eq!(output, std::path::PathBuf::from("./foo/picture.png"));

        let mut output = std::path::PathBuf::from("./foo");
        let args = crate::args::StandardArgs { 
            prefix: Some("nice".to_string()), 
            ..Default::default()
        };
        push_to_output(&args, &mut output, "picture", "png");
        assert_eq!(output, std::path::PathBuf::from("./foo/nice_picture.png"));
    }
}
