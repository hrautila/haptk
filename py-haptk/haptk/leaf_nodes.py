import itertools

def find_leaf_neighbors(hst):
    # A set of sibling nodes from the tree
    neighbor_sets = []

    for idx in hst.node_indices():
        # A node is leaf if it has no outgoing edges
        if len(hst.out_edges(idx)) == 0:

            # Checks if all siblings have exactly 1 sample
            is_paired_leaf = True

            # Add the first sibling
            neighbors = [idx]

            # Get parents by moving up to adj_direction
            parents = hst.adj_direction(idx, True)

            # Iterate parent for siblings (I think this loops only once in HSTs)
            for parent in parents:
                for sibling in hst.adj_direction(parent, False):
                    # Dont re-add the first sibling
                    if sibling != idx:
                        data = hst.get_node_data(sibling)

                        # Dont add if there is more than 1 sample in the node
                        if len(data['indexes']) == 1:
                            neighbors.append(sibling)
                            pass
                        else:
                            # A sibling has more than 1 sample
                            is_paired_leaf = False

            if is_paired_leaf:
                neighbors.sort()
                neighbor_sets.append(neighbors)

                
    # Sort
    neighbor_sets.sort()

    # Dedup because looping all leaf nodes produces duplicate pairs
    neighbor_sets = list(k for k,_ in itertools.groupby(neighbor_sets))

    return neighbor_sets

def find_average_leaf_node_length(hst):
    leaf_nodes = get_leaf_nodes(hst)

    sum = 0

    for node_idx in leaf_nodes:
        node_data = hst.get_node_data(node_idx)

        # A saturating substraction just to be sure
        length = max(node_data['stop'] - node_data['start'], 0)
        sum += length

    return sum / len(leaf_nodes)

def get_leaf_nodes(hst):
    leaf_nodes = []

    for idx in hst.node_indices():
        if len(hst.out_edges(idx)) == 0:
            leaf_nodes.append(idx)

    return leaf_nodes

   
