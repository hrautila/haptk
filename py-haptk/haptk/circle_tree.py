from ete3 import TreeStyle, TextFace, CircleFace,  NodeStyle, add_face_to_node

from haptk import utils

def circle_tree_style():
    ts = TreeStyle()

    ts.branch_vertical_margin = 10
    ts.show_leaf_name = False
    ts.show_scale = False

    ts.mode = "c"
    ts.arc_start = 0 # 0 degrees = 15:00 o'clock
    ts.arc_span = 355
    ts.min_leaf_separation = 0
    ts.root_opening_factor = 0
    ts.optimal_scale_level = "mid"
    ts.allow_face_overlap = True

    return ts

def is_maj_up_to(hst, node, value):
    iter_node = node
    # print(f"checking {iter_node.name} for majority")
    for _ in range(0, value):
        if iter_node.name == "0":
            return True

        if not iter_node.up.name == "0":
            # print("input node:", iter_node.up)
            is_big = is_majority_node(hst, iter_node.up)
            if is_big == False:
                # print(f"run {_}: {iter_node.up.name} is not big")
                return False

        iter_node = iter_node.up
            
    return True

def is_majority_node(hst, node):
    parent = node.up
    input_node_data = hst.get_node_data(int(node.name))
    value = len(input_node_data["indexes"])
    # print(f"input node {node.name} has {value} indexes")

    siblings = {}
    for child in parent.children:
        data = hst.get_node_data(int(child.name))
        if not child.name == node.name:
            siblings[child.name] = len(data["indexes"])


    # print(value, siblings.values())
    for v in siblings.values():
        if value < v: 
            return False
  

def get_children_idx_amounts(hst, node):
    return [len(hst.get_node_data(c.name)["indexes"]) for c in node.children]

def draw_circle_tree(hst, t, output, samples, samples_to_tag, colors):
    style = NodeStyle()
    style["size"] = 0

    narea_color = "#DAF7A6"
    larea_color = "#FFB6C1"

    spent_parents = []
    spent_grand_parents = []

    for n in t.traverse():
        n.set_style(style)
        
        if not n.is_leaf():
            indexes = hst.get_node_indexes(n.name)
            label = len(indexes)
            F = TextFace(label, tight_text=True, penwidth=30, fsize=8)
            F.rotation = 270
            n.add_face(F, column=0)

            count = 0
            for i in get_children_idx_amounts(hst, n):
                if i > 80:
                    count += 1

            if count > 1:
                haplotype = hst.get_node_haplotype(n.name)
                F = TextFace(f"{len(haplotype)}", fsize=30, tight_text=True, fgcolor="red")
                # F.rotation = 270
                n.add_face(F, column=0)
                
                

        if n.is_leaf():
            if samples_to_tag == []:
                n.set_style(utils.return_node_style("#000", 0))
                n.img_style["bgcolor"] = "#FFF"

            for (color_i, sample_list) in enumerate(samples_to_tag):
                results = [i for i in indexes if samples[i] in sample_list]
                if results:
                    n.set_style(utils.return_node_style(colors[color_i], 0))
                    n.img_style["bgcolor"] = colors[color_i]

            if n.up:
                parent = n.up
                parent_indexes = hst.get_node_indexes(parent.name)

                if len(parent_indexes) == 2:
                    if parent in spent_parents:
                        if is_maj_up_to(hst, n, 14):
                            # print(f"{n.name} is_maj")

                            haplotype = hst.get_node_haplotype(n.name)
                            start_idx = hst.get_node_start_idx(n.name)
                            stop_idx = hst.get_node_stop_idx(n.name)
                            start = hst.coords[start_idx]
                            stop = hst.coords[stop_idx]
                            # print(start, stop)
                            bp = stop['pos'] - start['pos']
                            F = TextFace(f"{len(haplotype)}, {bp}", fsize=30, fgcolor="white")
                            # F.margin_left = 50
                            n.add_face(F, column=0)

                            # n.set_style(utils.return_node_style("white", 0))
                            # n.img_style["bgcolor"] = "white"
                    else:
                        spent_parents.append(parent)


    return t

