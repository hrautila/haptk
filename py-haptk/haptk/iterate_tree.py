from ete3 import TreeStyle, TextFace, CircleFace,  NodeStyle, add_face_to_node
from scipy.stats import fisher_exact

from haptk import utils

def iterate_tree_style():
    ts = TreeStyle()

    ts.branch_vertical_margin = 10
    ts.show_leaf_name = False
    ts.show_scale = False

    ts.mode = "c"
    ts.arc_start = 0 # 0 degrees = 15:00 o'clock
    ts.arc_span = 355
    ts.min_leaf_separation = 0
    ts.root_opening_factor = 0
    ts.optimal_scale_level = "mid"
    ts.allow_face_overlap = True

    return ts

def iterate_tree_inner(hst, t, samples_to_tag):
    style = NodeStyle()
    style["size"] = 0

    for n in t.traverse():
        n.set_style(style)
        
        indexes = hst.get_node_indexes(n.name)

        tot_cases, tot_ctrls = 0, 0
        n_cases, n_ctrls = [], []

        for (i, sample_list) in enumerate(samples_to_tag):
            if i == 0:
                tot_cases = len(sample_list)
                n_cases = [i for i in indexes if hst.samples[i] in sample_list]
            if i == 1:
                tot_ctrls = len(sample_list)
                n_ctrls = [i for i in indexes if hst.samples[i] in sample_list]

        n_cases = len(n_cases)
        n_ctrls = len(n_ctrls)
        # res = fisher_exact([[n_cases, n_ctrls], [tot_cases - n_cases, tot_ctrls - n_ctrls]], alternative="greater")
        res = fisher_exact([[n_cases, n_ctrls], [tot_cases - n_cases, tot_ctrls - n_ctrls]])
        
        print(f"{n.name},{res.pvalue},{n_cases},{n_ctrls},{tot_cases},{tot_ctrls}")
        if res.pvalue < 0.0005:
            label = len(indexes)
            F = TextFace(f"{res.pvalue:.4}", tight_text=True, penwidth=30, fsize=8, fgcolor='red')
            F.rotation = 270
            F.background.color = "#FFF"
            F.margin_right = 10
            n.add_face(F, column=0)

        # if n.is_leaf() and res.pvalue < 0.005:
        label = len(indexes)
        F = TextFace(label, tight_text=True, penwidth=30, fsize=8)
        F.rotation = 270

        mix = res.pvalue * 1000
        if mix > 1:
            mix = 1
        color = utils.color_fader('red', 'blue', mix)

        F.background.color = color
        n.add_face(F, column=0)

    return t
