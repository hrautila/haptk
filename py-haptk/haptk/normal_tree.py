from ete3 import TreeStyle, TextFace, NodeStyle

from haptk import utils

def normal_tree_style():
    ts = TreeStyle()
    ts.rotation = 90
    ts.show_leaf_name = False
    ts.show_scale = False
    ts.min_leaf_separation = 0
    ts.branch_vertical_margin = 5
    ts.optimal_scale_level = "full"
    ts.allow_face_overlap = True
    return ts


def draw_normal_tree(hst, t, samples_to_tag, colors, proportions, show_haplotype, n_markers, show_pos):
    style = NodeStyle()
    style["size"] = 0

    for n in t.traverse():
        n.set_style(style)

        node_data = hst.get_node_data(int(n.name))

        label = len(node_data["indexes"])

        if proportions:
            top_data = hst.get_node_data(int('0'))
            samplen = len(top_data["indexes"])
            prc = label / samplen
            label = f"N={label} ({prc:.2f})"
        else:
            label = f"{label}"

        if show_haplotype:
            label = f"N={len(node_data['indexes'])}\n{hst.haplotype_string(node_data, n_markers)}"

        if show_pos:
            label = f"N={len(node_data['indexes'])}"
            haplotype = hst.haplotype_string(node_data, n_markers)
            start = node_data['start']['pos']
            stop = node_data['stop']['pos']
            label = f"{label}\n{haplotype}\nstart: {start}\nstop: {stop}"

        F = TextFace(label, tight_text=True, penwidth=30)
        F.rotation = 270
        n.add_face(F, column=0)

        if n.is_leaf():
            if samples_to_tag == []:
                n.set_style(utils.return_node_style("#000", 1))

            for (color_i, sample_list) in enumerate(samples_to_tag):
                results = [i for i in node_data["indexes"] if hst.get_sample_name(i) in sample_list]
                if results:
                    n.set_style(utils.return_node_style(colors[color_i], 7 - color_i))


    return t

