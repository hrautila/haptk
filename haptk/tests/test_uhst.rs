mod common;

#[cfg(test)]
#[cfg(feature = "clap")]
mod test_uhst {
    use super::*;

    use color_eyre::Result;

    use std::io::Write;
    use std::path::PathBuf;

    use haptk::{
        args::{Selection, StandardArgs},
        read_vcf::read_vcf_to_matrix,
        structs::Coord,
        subcommands::uhst_shard::{self, LocDirection},
    };

    #[test]
    fn matrix_to_graph() -> Result<()> {
        let args = StandardArgs {
            file: PathBuf::from(common::TEST_VCF),
            ..Default::default()
        };
        let mut vcf = read_vcf_to_matrix(&args, "chr9", 32, None, None, false).unwrap();

        let coord = Coord {
            contig: String::from("chr9"),
            reference: String::from("G"),
            alt: String::from("T"),
            pos: 32,
        };

        let g = uhst_shard::construct_uhst(&mut vcf, &LocDirection::Left, &coord, 1, false)?;
        let mut f = std::fs::File::create("tests/results/test.dot").unwrap();
        f.write_all(format!("{}", petgraph::dot::Dot::new(&g)).as_bytes())
            .unwrap();
        Ok(())
    }

    //--- Binary tests

    #[test]
    fn uhst_all() {
        run_uhst(Selection::All);

        let res = std::fs::read_to_string("tests/results/uhst_mbah.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype.csv").unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_ref() {
        run_uhst(Selection::OnlyRefs);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_refs.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_refs.csv")
            .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_alt() {
        run_uhst(Selection::OnlyAlts);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_alts.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_alts.csv")
            .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_longest() {
        run_uhst(Selection::OnlyLongest);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_longest.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res =
            std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_longest.csv")
                .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    fn run_uhst(selection: Selection) {
        let args = common::clap_standard_args(selection);

        let cmd = haptk::clap::SubCommand::Uhst {
            args,
            log_and_verbosity: crate::common::silent_verbosity(),
            threads: 8,
            min_size: 1,
            publish: false,
            sharded: false,
        };
        haptk::clap::run_cmd(cmd).unwrap();
    }
}
