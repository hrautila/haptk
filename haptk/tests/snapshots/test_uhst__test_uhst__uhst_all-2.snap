---
source: haptk/tests/test_uhst.rs
expression: res
---
"contig,pos,ref,alt,gt\n"
